### ISSUE TITLE
Screen corruption with quickfix

### ISSUE THREAD

**Issue by KillTheMule** [neovim/neovim#11580]
Feel free to edit the title, not sure if that really describes what I'm seeing

- `nvim --version`:
```
NVIM v0.5.0-246-g486ddb6eb
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/tmp/neovim/build/config -I/tmp/neovim/src -I/tmp/neovim/.deps/usr/include -I/usr/include -I/tmp/neovim/build/src/nvim/auto -I/tmp/neovim/build/include
```

- `vim -u DEFAULTS` (version: 8.1.1486) behaves differently? Yes (pretty sure it doesn't matter :))
- Operating system/version: Gentoo Linux
- Terminal name/version: XTerm(337)
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NONE`

In the neovim repo:
```
nvim src/nvim/eval.c -u NONE
:vs
:sp
<C-w>l
:copen
<C-w>k
:make
```

After the screen has filled with message from `make`, press `<C-c>` to cancel and return. This does not seem to matter (e.g. when using rust's `cargo`, which would fill half the screen in my case, it's half the screen that gets corrupted).

### Actual behaviour
The non-active buffers that have been covered by the output seem not to redraw. Plus, some other things don't look right, like the vertical window separator. See the first screenshot of

https://imgur.com/a/g4TdcfX

Note that this is with my full config, but it reproduces with `-u NONE`, and I have a hard time making screenshots.

Note the 2nd screenshot showing how it should look like. Redrawing the screen, or moving the cursor through all windows, fixes things.

**Comment by blueyed** [comment]
@KillTheMule 
Thanks for the report already.  Would be great if you could git-bisect it.

**Comment by KillTheMule** [comment]
Ahh right, I'll do that, but it might take 1 or 2 days.

**Comment by lacygoill** [comment]
It's caused by https://github.com/neovim/neovim/commit/b51ba122c1edfb769e72c25c4f14f469f59f1b8e.
https://github.com/neovim/neovim/issues/11511 may be related (at least they are caused by the same commit).

**Comment by bfredl** [comment]
I will investigate the coming days, ping me lest I forget.

**Comment by justinmk** [comment]
`CTRL-L` is a workaround meanwhile.

**Comment by phmarek** [comment]
I just interrupted my usual thought processes and found out that `:make` causes my screen corruption... so quickfix as well.

**Comment by KillTheMule** [comment]
That's not true for me, using `:make` without the qf window open seems to behave normally for me.

**Comment by rafikdraoui** [comment]
Here's an issue I have that might be related to this. When I have several loaded buffers and the quickfix list open, and I then do `:ls` followed by `:b <num>` (without clearing the output of `:ls`, like what happens in the "classic" `nnoremap <leader>b :ls<cr>:b` mapping), then the quickfix list is still open, but the old output of the buffers list is drawn on top of it.

One way to reproduce:

```
$ cd /usr/share/nvim/runtime/after/ftplugins
$ nvim -u NONE a*.vim
```

Then in neovim:

```
:grep setlocal *
:copen
<c-w> w
:ls:b awk.vim
```

---


nvim --version: NVIM v0.5.0-329-g91bd1ddf3  (but can also reproduce with v0.4.3)
vim behaves differently? Yes, cannot reproduce in vim
OS: Arch Linux & macOS
Terminal: Kitty 0.15.1
$TERM: xterm-kitty

**Comment by bfredl** [comment]
Here is a quickfix (modulo off-by-one error, need to do some more testing):

```diff
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index a45145172..97893223a 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -370,6 +370,17 @@ int update_screen(int type)
           grid_clear_line(&default_grid, default_grid.line_offset[i],
                           Columns, false);
         }
+        FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+          if (wp->w_floating) {
+            continue;
+          }
+          if (W_ENDROW(wp) > valid) {
+            wp->w_redr_type = MAX(wp->w_redr_type, NOT_VALID);
+          }
+          if (W_ENDROW(wp) + wp->w_status_height > valid) {
+            wp->w_redr_status = true;
+          }
+        }
       }
       msg_grid_set_pos(Rows-p_ch, false);
       msg_grid_invalid = false;
```

Ideally/eventually the compositor should get "smart" enough so we can just `msg_grid_set_pos()` and nothing else here. (i e compositor does its own delta tracking/merging).

**Comment by lacygoill** [comment]
Can confirm; the patch fixes the issues #11511 and #11580.
